test/stratumTest.js

```diff
@@ -0,0 +1,211 @@
+    var server;
+    this.beforeEach(function(){
+        server = new stratum.Server(config);
+    });
+
+    this.afterEach(function(){
+        server.close();
+    });
+
+    var submitMessage = {
+        id: null,
+        method: 'mining.submit',
+        params: 'block'
+    };
+
+    function assertBanned(address){
+        var [banned, _] = server.isBanned(address);
+        assert(banned);
+    }
+
+    function assertNotBanned(address){
+        var bannedIps = Object.keys(server.bannedIPs);
+        assert(bannedIps.find(ip => ip === address) === undefined);
+    }
+
+    function setupClient(client, callback){
+        client.setEncoding('utf8');
+        client.setNoDelay(true);
+        client.connect(config.pool.port);
+
+        var buffer = '';
+        client.on('data', function(data){
+            buffer += data;
+            if (buffer.indexOf('\n') !== -1){
+                var messages = buffer.split('\n');
+                var remain = buffer.slice(-1) === '\n' ? '' : messages.pop();
+                messages.forEach(message => {
+                    if (message === '') return;
+                    callback(JSON.parse(message));
+                });
+                buffer = remain;
+            }
+        });
+    }
+
+    it('should work as expected', function(done){
+        var client = net.Socket();
+        var jobs = [new DummyJob()];
+
+        setupClient(client, function(message){
+            switch(message.method) {
+                case 'mining.set_difficulty':
+                    expect(message.params).to.deep.equal([config.pool.diff]);
+                    break;
+                case 'mining.notify':
+                    expect(message.params).to.deep.equal(jobs.map(job => job.getJobParams()));
+                    break;
+                case 'mining.submit_result':
+                    expect(message.result).equal(true);
+                    client.end();
+                    done();
+                    break;
+                default:
+                    assert(false, 'Unknow message type: ' + message.method);
+            }
+        });
+
+        client.on('connect', function(){
+            var clientIds = Object.keys(server.stratumClients);
+            var stratumClient = server.stratumClients[clientIds[0]];
+            stratumClient.on('submit', function(params, callback){
+                expect(params).equal(submitMessage.params);
+                callback(null, true);
+            });
+
+            assert(clientIds.length === 1);
+            server.broadcastMiningJobs(jobs);
+            client.write(JSON.stringify(submitMessage) + '\n');
+        });
+    })
+
+    it('should disconnect if client is banned', function(done){
+        var client = net.Socket();
+        client.connect(config.pool.port);
+        client.on('connect', function(){
+            server.addBannedIP(client.remoteAddress);
+            assertBanned(client.remoteAddress);
```

Comment:
Is it possible to check the handlers or caches for client (in hash table) are closed in `server` ?